base/abstractarray.jl

```diff
@@ -889,9 +849,50 @@ function typed_hvcat(T::Type, rows::Tuple{Vararg{Int}}, xs::Number...)
     hvcat_fill(Array{T}(nr, nc), xs)
 end
 
-function hvcat(rows::Tuple{Vararg{Int}}, xs::Number...)
-    T = promote_typeof(xs...)
-    typed_hvcat(T, rows, xs...)
+hvcat(rows::Tuple{Vararg{Int}}, xs::AbstractMatrix...) = typed_hvcat(promote_eltype(xs...), rows, xs...)
+hvcat{T}(rows::Tuple{Vararg{Int}}, xs::AbstractMatrix{T}...) = typed_hvcat(T, rows, xs...)
+
+function typed_hvcat{T}(TT::Type{T}, rows::Tuple{Vararg{Int}}, as::AbstractMatrix{T}...)
```

Comment:
This requires all the argument matrices to have the output type `T` as their element type already.

Also could you keep the code in place above to get a smaller diff?